articles/external_qos_file.md

```diff
@@ -0,0 +1,658 @@
+  The complexity of this approach doesn't seem to be worth it.
+Note 2:
+  The rmw specific payload could simply be an string, instead of key value pairs.
+  In that case, the `profile_name` sub-tag would be deleted and `my_qos_profile_name` would be the content of the string.
+
+## File format
+
+As showed in the above examples, both YAML or XML can correctly represent the desired file format.
+
+XML advantages:
+- Can be easily verified.
+
+YAML advantages:
+- It can be argued that the format is simpler and more readable than XML.
+- The format proposed in the above examples can easily be integrated with the current parameter file format.
+  That integration would lead in a unique "configuration file", that could be further extended in the future.
+  For example, passing remapping rules in it.
+
+## Explicitly allowing external configurability in entities
+
+The node options could have a flag to ignore the new `--qos-file` argument.
+In a similar way, there could be an option for publishers/subscriptions/clients and services to allow loading the QoS from an external source (side load).
+
+In nodes designed to be reused, it does make sense to allow side-loading the qos settings of all entities.
+To make this use case easier, the following options can be added:
+
+- `rcl_node_options_t` is extended with a `bool allow_side_loading_qos`.
+- The following enum is defined:
+  ```c
+  typedef enum rcl_side_load_qos_t {
+      RCL_SIDE_LOAD_QOS_DEFAULT;
+      RCL_SIDE_LOAD_QOS_ENABLED;
+      RCL_SIDE_LOAD_QOS_DISABLED;
+  } rcl_side_load_qos_t;
+  ```
+- A `rcl_side_load_qos_t allow_side_loading_qos` member is added to publisher/subscription/client/service options, it will copy the behavior defined in the `rcl_node_options_t` of its parent node when the enum value is `RCL_SIDE_LOAD_QOS_DEFAULT`.
+
+## Which QoS policies can be externally modified?
+
+It might not make sense to be able to externally modify some of the QoS policies.
+For example, modifying `rmw_qos_liveliness_policy_t` externally to `MANUAL_BY_TOPIC`.
+
+### Analysis for each policy
+
+#### History kind, history depth, durability, reliability, lifespan
+
+These four policies provide an abstraction in a way that the code to publish or take a message from a publisher/subscription doesn't have to be changed if the history kind or history depth changed.
+Thus, it does not make sense to restrict the external configurability of them.
+
+#### Deadline
+
+If the node is not listening to "deadline missed" events for that entity, modifying it externally won't have any effect.
+For nodes that are using this feature, allowing to externally configure it makes sense.
+
+#### Liveliness and liveliness_lease_duration
+
+Different values of `liveliness_lease_duration` always make sense, regardless of how the node is implemented.
+In the case of the liveliness kind, `MANUAL_BY_TOPIC` policy does not make sense if the author of the node doesn't thought about it.
```

Comment:
```suggestion
In the case of the liveliness kind, `MANUAL_BY_TOPIC`, it does not make sense if the node author hasn't considered it.
```

I'm a bit confused by this section. Are we saying that we will never allow externally setting a value of `MANUAL_BY_TOPIC`?

Reply:
> I'm a bit confused by this section. Are we saying that we will never allow externally setting a value of MANUAL_BY_TOPIC?

I'm saying that it almost never makes sense, except the author considered both cases which is strange.
That's why I proposed that the default QoS validation callback allows changing everything except the liveliness QoS policy.